### ISSUE TITLE
Refutation not working

### ISSUE THREAD

**Issue by imrryr** [memphis-iis/mofacts-ies#455]
**Describe the bug**
Not working

![mofacts_app and 14 more pages - Personal - Microsoft​ Edge 12_5_2021 3_56_16 PM](https://user-images.githubusercontent.com/13023147/144765647-e4b1ce43-1edc-4b21-b29d-7ea6d9354739.png)
![mofacts_app and 14 more pages - Personal - Microsoft​ Edge 12_5_2021 3_56_11 PM](https://user-images.githubusercontent.com/13023147/144765648-02cfe982-bc22-44dc-870b-dd5c0f1e9f7e.png)

**To Reproduce**
Run posted tdf

**Expected behavior**
e.

**Comment by MegaGeese** [comment]
This appears to be an issue with the answer associated with a stim. Looking through the file `2021-01-17-glossary-machine-dictionary-tagged.json` the entry "brainstem" exists, but the entry for the correct answer "impulses" is not. Refutational feedback only works when both the incorrect answer and the correct answer are in the glossary file. I've found that a lot of the issues are being caused by the correct answer being plural in the stim but singular in the glossary file.

**Comment by imrryr** [comment]
Adding Andrew Olney to help us figure this out. 

The hints listed should always trigger refutations, otherwise our procedure to list the hints has a bug.

**Comment by MegaGeese** [comment]
From what I found in my debugging, the refutational feedback requires that both the incorrect and correct answers be in the glossary file exactly as they appear in the stim. Many of the answers in the chapter 10 stim do not appear in the glossary file, they are either not in there or they are plural in the stim and singular in the glossary. 

For example, in the second picture the correct answer in the stim is "impulses" but in the glossary it is written as "impulse"

**Comment by imrryr** [comment]
OK, this is great to know, since we should probably ignore the 's. However, that doesn't explain to me why the list of hints is wrong. The hints are by definition only supposed to be provided WHEN they trigger a refutation. So how do we get hints if the answer and the wrong answer are not in the glossary? You see, we should not. The whole purpose of the list of hints is to identify when refutation will trigger. Does code generate the hints, were they provided by Andrew? @aolney

**Comment by MegaGeese** [comment]
The list of words is created when mofacts is deployed using the file `2021-02-25-elaboratedFeedbackCache.json`. Inside of that file it has relations between correct and incorrect words that are used to generate the teacher aid. 

In the 2nd screenshot the answer is impulses, impulses is listed inside elaboratedFeedbackCache and so is impulse; however, since only impulse is listed in the glossary file it does not generate refutational feedback for impulses. 

In short, mofacts uses the elaboratedFeedbackCache file to generate the words that should trigger refutation based on the answer, and the glossary file is used to generate this refutational feedback. If there is a word that is in the elaboratedFeedbackCache file that is not in glossary it will cause the issue posted in your second screenshot.

What I can do is add a check when generating that list that looks in the glossary file and only adds words to the list if they exist in the glossary file.

**Comment by imrryr** [comment]
So, both those files come from @aolney correct? I'd prefer to get the corrected files if we can.

**Comment by MegaGeese** [comment]
Rusty and I got the files from Andrew Tackett. I'm not sure if he got them from Olney.

**Comment by imrryr** [comment]
We'll just sit tight on this issue until he has time for us

**Comment by aolney** [comment]
What should be happening is this:

- Look up on elaborated feedback cache. If the correct/incorrect appear there, use it (this is the deep learning stuff, cached)
- If the cache hit fails, look up in glossary. If the correct/incorrect appear there, use their definitions to construct feedback

The most recent elaborated feedback cache is from 7/17/21. I sent it to Tackett as an email attachment.

The debugging info shown to testers (i.e. use these words and you will trigger an elaborated feedback) is not generated by my code. I assume Tackett set it up using the two sources of feedback above.

**Comment by imrryr** [comment]
@aolney Did you read the whole thread though? It is triggered by the lack of plurals in your files I think. Can we handle that by adding plurals to the files? Plurals can be hard to infer...

**Comment by imrryr** [comment]
@MegaGeese it sounds like we need to fix Andrew T's code to generate the hints as one action here.

**Comment by imrryr** [comment]
@aolney did you read Rhodes comment? He says the mappings are in your file....

**Comment by imrryr** [comment]
@aolney @MegaGeese Hard to figure this issue out. 

Is it possible that you map things in the cache file Andrew, and then objects that are mapped are not in the glossary?

**Comment by MegaGeese** [comment]
Looking at what @aolney said, it looks like with how it is written, it is not checking the elaborated feedback cache at all and going straight to the glossary. I can work to fix that.

**Comment by MegaGeese** [comment]
to elaborate, when I say 'it' I mean the code in methods.js that gets the feedback from the cache.

**Comment by aolney** [comment]
Here are some links to the code:

- [CachedElaboratedFeedback](https://github.com/aolney/mofacts-automated-authoring/blob/master/src/CachedElaboratedFeedback.fs)
- [DefinitionalFeedback](https://github.com/aolney/mofacts-automated-authoring/blob/master/src/DefinitionalFeedback.fs)

In DefinitionalFeedback I previously used spelling correction to recover student answers that were misspelled, but we took that out because it used too much RAM. If we are OK with using RAM, I can look at other options for spelling correction that are more memory efficient. Note this would likely add an NPM dependency because I'd be using a 3rd party library. I assume that's OK.

Currently there is no handling of plurals. The cache has both singular and plural forms - whatever was in the glossary is what we used. I could add in some code to query these variations using regular plural rules, e.g. "s" and "es". 

While I've been writing this, there are been more comments suggesting that the current code is going straight to definitions. I can think of 2 ways this could happen:

- `CachedElaboratedFeedback.GenerateFeedback` is not being called
- It is being called, but the cache file is not accessible, in which case it automatically rolls off to definitional feedback
